modules/tinymce/src/core/main/ts/init/LicenseKeyManager.ts

```diff
@@ -0,0 +1,133 @@
+    const { plugin } = data || {};
+    // Premium plugins are not allowed if 'gpl' is given as the license_key
+    if (Type.isString(plugin)) {
+      return Promise.resolve(false);
+    } else {
+      return Promise.resolve(true);
+    }
+  },
+};
+
+const ADDON_KEY = 'commercial';
+const PLUGIN_CODE = 'licensekeymanager';
+
+const setup = (): LicenseKeyManagerLoader => {
+  const addOnManager = AddOnManager<LicenseKeyManager>();
+
+  // TODO: TINY-12081: Add integrity/checksum check
+  const add = async (addOn: LicenseKeyManagerAddon): Promise<void> => {
+    addOnManager.add(ADDON_KEY, addOn);
+  };
+
+  const load = (editor: Editor, suffix: string): void => {
+    const licenseKey = Options.getLicenseKey(editor);
+    if (licenseKey !== 'gpl' && !Obj.has(addOnManager.urls, ADDON_KEY)) {
+      const url = `plugins/${PLUGIN_CODE}/plugin${suffix}.js`;
+
+      addOnManager.load(ADDON_KEY, url).catch(() => {
+        // eslint-disable-next-line no-console
+        console.error(`Failed to load license key manager from url ${url}`);
+        displayNotification(editor, 'Failed to load license key manager');
+      });
+    }
+  };
+
+  const isLoaded = (editor: Editor) => {
+    const licenseKey = Options.getLicenseKey(editor);
+    if (licenseKey === 'gpl') {
+      return true;
+    } else {
+      return Type.isNonNullable(addOnManager.get(ADDON_KEY));
+    }
+  };
+
+  const init = (editor: Editor): void => {
+    const lockLicenseKeyManager = () => {
+      Object.defineProperty(editor, 'licenseKeyManager', {
+        writable: false,
+        configurable: false,
+        enumerable: true,
+      });
+    };
+
+    const licenseKey = Options.getLicenseKey(editor);
+    if (licenseKey === 'gpl') {
+      editor.licenseKeyManager = GplLicenseKeyManager;
+      lockLicenseKeyManager();
+      return;
+    }
+
+    // CommercialLicenseKeyManager should be nonnullable here as the
+    // editor will not load without a license key manager constructor
+    const CommercialLicenseKeyManager = addOnManager.get(ADDON_KEY);
+    if (Type.isNullable(CommercialLicenseKeyManager)) {
+      return;
+    }
+
+    const commercialLicenseKeyManagerApi = CommercialLicenseKeyManager(editor, addOnManager.urls[ADDON_KEY]);
+    editor.licenseKeyManager = commercialLicenseKeyManagerApi || {};
+    lockLicenseKeyManager();
+    const verify = editor.licenseKeyManager.verify;
+    const validate = editor.licenseKeyManager.validate;
+    if (!Type.isFunction(verify) || !Type.isFunction(validate)) {
+      // eslint-disable-next-line no-console
+      console.error(`Incorrect license key manager API`);
+      displayNotification(editor, 'Failed to intialize license key manager');
+      ForceReadonly.forceReadonly(editor);
+      return;
+    }
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    verify(editor);
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    validate(editor);
```

Comment:
Why is this called here? Just as a stop gap?

Reply:
Thinking about this more, I might remove the `verify` and `validate` calls here and move them into the licensekeymanager plugin code

Reply:
I have added a comment in the code for clarification
```
// Validation of the license key is done asynchronously and does
// not block initialization of the editor
// The validate function is expected to set the editor to the correct
// state depending on if the license key is valid or not
```